File: app/src/main/java/com/simpleattendance/HistoryActivity.java
package com.simpleattendance;

import android.content.Intent;
import android.database.Cursor;
import android.os.Bundle;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.button.MaterialButton;

import java.util.ArrayList;
import java.util.List;

public class HistoryActivity extends AppCompatActivity {

    private DatabaseHelper databaseHelper;
    private TextView totalClassesCount, totalStudentsCount;
    private Spinner classFilterSpinner;
    private RecyclerView historyRecyclerView;
    private LinearLayout emptyStateLayout;
    
    private List<ClassModel> classList;
    private List<AttendanceSession> sessionList;
    private HistorySessionAdapter historyAdapter;
    private int selectedClassId = -1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_history);

        databaseHelper = new DatabaseHelper(this);
        initViews();
        setupToolbar();
        loadData();
        setupSpinner();
        setupRecyclerView();
        updateSummaryStats();
    }

    private void initViews() {
        totalClassesCount = findViewById(R.id.totalClassesCount);
        totalStudentsCount = findViewById(R.id.totalStudentsCount);
        classFilterSpinner = findViewById(R.id.classFilterSpinner);
        historyRecyclerView = findViewById(R.id.historyRecyclerView);
        emptyStateLayout = findViewById(R.id.emptyStateLayout);
    }

    private void setupToolbar() {
        ImageView backButton = findViewById(R.id.backButton);
        backButton.setOnClickListener(v -> onBackPressed());
    }

    private void loadData() {
        classList = new ArrayList<>();
        sessionList = new ArrayList<>();
        
        // Load all classes
        classList = databaseHelper.getAllClasses();
    }

    private void setupSpinner() {
        List<String> classNames = new ArrayList<>();
        classNames.add("All Classes");
        
        for (ClassModel classModel : classList) {
            String displayName = classModel.getBranch() + " - " + 
                               classModel.getSemester() + " " + classModel.getSection();
            if (classModel.getSubject() != null && !classModel.getSubject().isEmpty()) {
                displayName += " (" + classModel.getSubject() + ")";
            }
            classNames.add(displayName);
        }

        ArrayAdapter<String> adapter = new ArrayAdapter<>(this, 
            android.R.layout.simple_spinner_item, classNames);
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        classFilterSpinner.setAdapter(adapter);

        classFilterSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                if (position == 0) {
                    selectedClassId = -1; // All classes
                } else {
                    selectedClassId = classList.get(position - 1).getId();
                }
                loadAttendanceSessions();
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {}
        });
    }

    private void setupRecyclerView() {
        historyAdapter = new HistorySessionAdapter(sessionList, new HistorySessionAdapter.OnSessionClickListener() {
            @Override
            public void onSessionClick(AttendanceSession session) {
                openSessionReport(session);
            }
        });
        historyRecyclerView.setLayoutManager(new LinearLayoutManager(this));
        historyRecyclerView.setAdapter(historyAdapter);
    }

    private void loadAttendanceSessions() {
        sessionList.clear();
        
        // Load actual attendance sessions from database
        if (selectedClassId == -1) {
            // Load all attendance sessions
            sessionList.addAll(databaseHelper.getAllAttendanceSessions());
        } else {
            // Load sessions for specific class
            sessionList.addAll(databaseHelper.getAttendanceSessionsByClass(selectedClassId));
        }
        
        if (sessionList.isEmpty()) {
            historyRecyclerView.setVisibility(View.GONE);
            emptyStateLayout.setVisibility(View.VISIBLE);
        } else {
            historyRecyclerView.setVisibility(View.VISIBLE);
            emptyStateLayout.setVisibility(View.GONE);
        }
        
        historyAdapter.notifyDataSetChanged();
        updateSummaryStats();
    }

    private void updateSummaryStats() {
        totalClassesCount.setText(String.valueOf(classList.size()));
        
        // Count total students across all classes
        int totalStudents = 0;
        for (ClassModel classModel : classList) {
            Cursor studentCursor = databaseHelper.getStudentsCursorByClass(classModel.getId());
            totalStudents += studentCursor.getCount();
            studentCursor.close();
        }
        totalStudentsCount.setText(String.valueOf(totalStudents));
    }

    private void openSessionReport(AttendanceSession session) {
        // Navigate to report activity for this specific session
        AttendanceReport report = databaseHelper.generateReport(session.getId());
        
        Intent intent = new Intent(this, ReportActivity.class);
        intent.putExtra("CLASS_NAME", session.getClassName());
        intent.putExtra("SUBJECT_NAME", "General");
        intent.putExtra("SESSION_DATE", session.getDate());
        intent.putExtra("SESSION_TIME", session.getTime());
        intent.putExtra("PRESENT_COUNT", report.getPresentCount());
        intent.putExtra("ABSENT_COUNT", report.getAbsentCount());
        intent.putStringArrayListExtra("ABSENT_STUDENTS", new ArrayList<>(report.getAbsentStudents()));
        startActivity(intent);
    }
}

File: app/src/main/java/com/simpleattendance/ui/history/HistoryActivity.kt
package com.simpleattendance.ui.history

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.simpleattendance.data.local.entity.AttendanceSessionEntity
import com.simpleattendance.databinding.ActivityHistoryBinding
import com.simpleattendance.ui.main.MainActivity
import com.simpleattendance.ui.report.ReportActivity
import com.simpleattendance.util.HapticUtils
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import javax.inject.Inject

@AndroidEntryPoint
class HistoryActivity : AppCompatActivity() {
    
    private lateinit var binding: ActivityHistoryBinding
    private val viewModel: HistoryViewModel by viewModels()
    
    @Inject
    lateinit var hapticUtils: HapticUtils
    
    private val historyAdapter by lazy {
        HistoryAdapter(
            onSessionClick = { session ->
                hapticUtils.lightTap()
                openReport(session)
            },
            onSessionLongClick = { session ->
                hapticUtils.mediumImpact()
                confirmDelete(session)
            }
        )
    }
    
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityHistoryBinding.inflate(layoutInflater)
        setContentView(binding.root)
        
        setupToolbar()
        setupRecyclerView()
        setupFilter()
        setupBottomNavigation()
        observeState()
    }
    
    private fun setupToolbar() {
        binding.toolbar.setNavigationOnClickListener {
            hapticUtils.lightTap()
            finish()
        }
    }
    
    private fun setupRecyclerView() {
        binding.recyclerView.apply {
            layoutManager = LinearLayoutManager(this@HistoryActivity)
            adapter = historyAdapter
            setHasFixedSize(true)
        }
    }
    
    private fun setupFilter() {
        // Will be populated when classes are loaded
    }
    
    private fun setupBottomNavigation() {
        // Bottom navigation removed from layout - navigation handled by toolbar
    }
    
    private fun observeState() {
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.uiState.collect { state ->
                    binding.emptyState.visibility = if (state.isEmpty && !state.isLoading) View.VISIBLE else View.GONE
                    binding.recyclerView.visibility = if (!state.isEmpty && !state.isLoading) View.VISIBLE else View.GONE
                    
                    // Update filter spinner
                    if (state.classes.isNotEmpty()) {
                        val classNames = listOf("All Classes") + state.classes.map { it.fullDisplayName }
                        val adapter = ArrayAdapter(this@HistoryActivity, android.R.layout.simple_spinner_item, classNames)
                        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
                        binding.classFilterSpinner.adapter = adapter
                        
                        binding.classFilterSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
                            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                                hapticUtils.lightTap()
                                val classId = if (position == 0) null else state.classes[position - 1].id
                                viewModel.filterByClass(classId)
                            }
                            override fun onNothingSelected(parent: AdapterView<*>?) {}
                        }
                    }
                    
                    historyAdapter.submitList(state.sessions)
                }
            }
        }
    }
    
    private fun openReport(session: AttendanceSessionEntity) {
        val intent = Intent(this, ReportActivity::class.java)
        intent.putExtra("sessionId", session.id)
        startActivity(intent)
    }
    
    private fun confirmDelete(session: AttendanceSessionEntity) {
        MaterialAlertDialogBuilder(this)
            .setTitle("Delete Session")
            .setMessage("Are you sure you want to delete this attendance session?")
            .setPositiveButton("Delete") { _, _ ->
                hapticUtils.mediumImpact()
                viewModel.deleteSession(session)
            }
            .setNegativeButton("Cancel", null)
            .show()
    }
}
